tests/integrated/test_suite

```diff
@@ -100,42 +206,80 @@ test_type = "Making" if args.make else "Running"
 
 print("======= {} {} {} tests ========".format(test_type, len(tests), savepath.split("/")[-1]))
 
+avail = num_jobs
+tester = sorted([Test(t, command) for t in tests],
+                key=lambda x: x.cost, reverse=True)
+cost_remain = sum(x.cost for x in tester)
+if js_read:
+    js_tokens = os.read(js_read,cost_remain-1)
+    num_jobs = len(js_tokens)+1
+
+torun=[i for i in range(len(tests))]
+running=[]
+
 longest = max([len(s) for s in tests])
 output = '%-'+str(longest+1)+'s'
-for t in tests:
-    os.chdir(t)
-
-    print(output % t, end='')
-    sys.stdout.flush()  # Make sure name is flushed
-
-    # Check requirements
-    req_met, req_expr = requirements.check("./runtest")
-    if not req_met:
-        print("S - {0} => False".format(req_expr))
-        sys.stdout.flush()
-        os.chdir(savepath)
-        continue
-
-    start_time_ = time.time()
-    # Run test, piping stdout so it is not sent to console
-    status, out = shell(command, pipe=True)
-
-    if status != 0:
-        # ❌ Failed
-        print(u"\u274C", end='')  # No newline
-        failed.append((t, out))
+
+while len(torun)+len(running):
+    if avail:
+        # See whether we can start one
+        for i in torun:
+            if tester[i].cost <= avail:
+                torun.remove(i)
+                job = tester[i]
+                job.start()
+                running.append(job)
+                avail -= job.cost
+                cost_remain -= job.cost
+        # if nothing is running, start the first
+        if running == []:
+            tostart = torun.pop(0)
+            job = tester[tostart]
+            job.start()
+            running.append(job)
+            avail -= job.cost
+            cost_remain -= job.cost
+
+    # Are any jobs finished
+    for job in running:
+        job.join(.1)
+        if not job.is_alive():
+            if job.status:
+                failed.append([job.name,job.output])
+            avail += job.cost
+            running.remove(job)
+
+    # Try to get more jobs
+    if torun:
+        if js_read is not None:
+            try:
+                newt = os.read(js_read,cost_remain)
```

Comment:
```suggestion
                new_test = os.read(js_read,cost_remain)
```

I might just be too tired, but I can't read this and not think of [newts](https://en.wikipedia.org/wiki/Newt)

Reply:
It should be new_tokens :-)
I've updated and learned a new word, thanks